File: packages/typescript/src/constituencies.ts
import type Database from "better-sqlite3";
import { Counties } from "./counties.js";
import { EntityNotFoundError } from "./errors.js";
import type { Constituency } from "./types.js";

interface ConstituencyRow {
  id: number;
  code: string;
  name: string;
  county_id: number;
  status: string;
}

function toConstituency(row: ConstituencyRow): Constituency {
  return { id: row.id, code: row.code, name: row.name, countyId: row.county_id, status: row.status };
}

export class Constituencies {
  constructor(private readonly db: Database.Database) {}

  list(): Constituency[] {
    const rows = this.db
      .prepare("SELECT id, code, name, county_id, status FROM constituencies ORDER BY name")
      .all() as ConstituencyRow[];
    return rows.map(toConstituency);
  }

  get(nameOrCode: string): Constituency {
    const row = this.db
      .prepare(
        "SELECT id, code, name, county_id, status FROM constituencies " +
          "WHERE name = ? COLLATE NOCASE OR code = ? COLLATE NOCASE",
      )
      .get(nameOrCode, nameOrCode) as ConstituencyRow | undefined;
    if (!row) throw new EntityNotFoundError("constituency", nameOrCode);
    return toConstituency(row);
  }

  inCounty(countyNameOrCode: string): Constituency[] {
    const county = new Counties(this.db).get(countyNameOrCode);
    const rows = this.db
      .prepare(
        "SELECT id, code, name, county_id, status FROM constituencies " +
          "WHERE county_id = ? ORDER BY name",
      )
      .all(county.id) as ConstituencyRow[];
    return rows.map(toConstituency);
  }
}

File: packages/python/src/kenya_data/constituencies.py
"""Constituency lookups."""
from __future__ import annotations

import builtins
import sqlite3

from .counties import Counties
from .exceptions import EntityNotFoundError
from .models import Constituency


class Constituencies:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def list(self) -> builtins.list[Constituency]:
        rows = self._conn.execute(
            "SELECT id, code, name, county_id, status FROM constituencies ORDER BY name"
        ).fetchall()
        return [Constituency(**dict(r)) for r in rows]

    def get(self, name_or_code: str) -> Constituency:
        row = self._conn.execute(
            "SELECT id, code, name, county_id, status FROM constituencies "
            "WHERE name = ? COLLATE NOCASE OR code = ? COLLATE NOCASE",
            (name_or_code, name_or_code),
        ).fetchone()
        if row is None:
            raise EntityNotFoundError("constituency", name_or_code)
        return Constituency(**dict(row))

    def in_county(self, county_name_or_code: str) -> builtins.list[Constituency]:
        county = Counties(self._conn).get(county_name_or_code)
        rows = self._conn.execute(
            "SELECT id, code, name, county_id, status FROM constituencies "
            "WHERE county_id = ? ORDER BY name",
            (county.id,),
        ).fetchall()
        return [Constituency(**dict(r)) for r in rows]
